vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerRequestImpl.java

```diff
@@ -91,4 +104,54 @@ public GrpcServerResponse<Req, Resp> response() {
   public HttpConnection connection() {
     return httpRequest.connection();
   }
+
+  @Override
+  public void handle(Buffer chunk) {
+    if (notGrpcWebText()) {
+      super.handle(chunk);
+      return;
+    }
+    if (grpcWebTextBuffer == EMPTY_BUFFER) {
+      if (divisibleByFour(chunk.length())) {
+        super.handle(Buffer.buffer(DECODER.decode(chunk.getBytes())));
+      } else {
+        grpcWebTextBuffer = chunk.copy();
+      }
+      return;
+    }
+    bufferAndDecode(chunk);
+  }
+
+  private boolean notGrpcWebText() {
+    return grpcWebTextBuffer == null;
+  }
+
+  private boolean divisibleByFour(int len) {
+    return (len & 0b11) == 0;
+  }
+
+  private void bufferAndDecode(Buffer chunk) {
+    grpcWebTextBuffer.appendBuffer(chunk);
+    int len = grpcWebTextBuffer.length();
+    // Decode base64 content as soon as we have more bytes than a multiple of four.
+    // We could instead wait for the buffer length to be a multiple of four,
+    // But then in the worst case we may have to buffer the whole request.
+    int maxBase64 = maxBase64(len);
+    if (maxBase64 == len) {
+      Buffer decoded = Buffer.buffer(DECODER.decode(grpcWebTextBuffer.getBytes()));
+      grpcWebTextBuffer = EMPTY_BUFFER;
+      super.handle(decoded);
+    } else if (maxBase64 > 0) {
+      Buffer decoded = Buffer.buffer(DECODER.decode(grpcWebTextBuffer.getBytes(0, maxBase64)));
+      grpcWebTextBuffer = grpcWebTextBuffer.getBuffer(maxBase64, len);
+      super.handle(decoded);
+    }
+  }
+
+  private int maxBase64(int len) {
+    while (!divisibleByFour(len)) {
```

Comment:
isn't that `len & 0x3` ?

Reply:
Yes, that's what the `divisibleByFour` method  above does. It's a small private method, its content should be easily inlined by the JVM.

`divisibleByFour` is used by another method as well, and it makes the code more readable.